File: backend/app/schemas/team.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.database.models.owner_profile import OwnerType
from app.database.models.property import PropertyStatus, PropertyType
from app.schemas.property import PropertyResponse


class TeamOwnerResponse(BaseModel):
    owner_profile_id: int
    user_id: UUID
    owner_type: OwnerType
    display_name: str
    company_name: str | None = None
    first_name: str
    last_name: str
    # None cuando el alta se hizo sin email real (dirección provisional
    # @coflow.pending): no es un contacto utilizable.
    email: str | None = None
    phone: str | None = None
    # True si ya puede entrar: activó su enlace (contraseña) o usa Google.
    account_activated: bool
    # Solo en el buscador de clientes; en listados de viviendas no aplica.
    property_count: int | None = None


class TeamPropertySummaryResponse(BaseModel):
    id: int
    title: str
    status: PropertyStatus
    property_type: PropertyType
    address_line: str
    city: str
    neighborhood: str | None = None
    postal_code: str
    total_monthly_rent: int | None = None
    deposit: int | None = None
    bedrooms: int
    bathrooms: int
    max_tenants: int
    surface_m2: int | None = None
    available_from: date | None = None
    cover_image_url: str | None = None
    image_count: int
    missing_fields: list[str] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime
    ready_at: datetime | None = None
    owner: TeamOwnerResponse


class TeamPropertyDetailResponse(PropertyResponse):
    owner: TeamOwnerResponse
    missing_fields: list[str] = Field(default_factory=list)


class TeamPropertyUpdate(BaseModel):
    # Mismas reglas relajadas que AssistedPropertyCreate: el equipo guarda
    # borradores a medias. Los mínimos reales se exigen al publicar
    # (PropertyService._missing_ready_fields). Un None en una columna
    # obligatoria significa "no tocar"; en una opcional, "vaciar".
    title: str | None = Field(default=None, max_length=150)
    description: str | None = None
    property_type: PropertyType | None = None

    address_line: str | None = Field(default=None, max_length=200)
    city: str | None = Field(default=None, max_length=100)
    province: str | None = Field(default=None, max_length=100)
    postal_code: str | None = Field(default=None, max_length=15)
    neighborhood: str | None = Field(default=None, max_length=120)

    latitude: float | None = None
    longitude: float | None = None

    surface_m2: int | None = Field(default=None, gt=0)
    bedrooms: int | None = Field(default=None, ge=0)
    bathrooms: int | None = Field(default=None, ge=1)
    floor: str | None = Field(default=None, max_length=20)
    has_elevator: bool | None = None
    furnished: bool | None = None
    max_tenants: int | None = Field(default=None, ge=1)

    total_monthly_rent: int | None = Field(default=None, ge=0)
    deposit: int | None = Field(default=None, ge=0)
    utilities_included: bool | None = None

    available_from: date | None = None
    minimum_stay_months: int | None = Field(default=None, ge=1)

    pets_allowed: bool | None = None
    smoking_allowed: bool | None = None
    couples_allowed: bool | None = None
    students_allowed: bool | None = None
    registration_allowed: bool | None = None
    additional_requirements: str | None = Field(default=None, max_length=2000)

    amenity_ids: list[int] | None = None

File: frontend/src/types/team.ts
import type { Property, PropertyStatus, PropertyType } from "./property";

export type OwnerType = "INDIVIDUAL" | "COMPANY" | "AGENCY";

export interface TeamOwner {
  owner_profile_id: number;
  user_id: string;
  owner_type: OwnerType;
  display_name: string;
  company_name: string | null;
  first_name: string;
  last_name: string;
  /** null si el alta se hizo sin email real. */
  email: string | null;
  phone: string | null;
  account_activated: boolean;
  /** Solo en el buscador de clientes. */
  property_count: number | null;
}

export interface TeamPropertySummary {
  id: number;
  title: string;
  status: PropertyStatus;
  property_type: PropertyType;
  address_line: string;
  city: string;
  neighborhood: string | null;
  postal_code: string;
  total_monthly_rent: number | null;
  deposit: number | null;
  bedrooms: number;
  bathrooms: number;
  max_tenants: number;
  surface_m2: number | null;
  available_from: string | null;
  cover_image_url: string | null;
  image_count: number;
  missing_fields: string[];
  created_at: string;
  updated_at: string;
  ready_at: string | null;
  owner: TeamOwner;
}

export interface TeamPropertyDetail extends Property {
  owner: TeamOwner;
  missing_fields: string[];
}

/** Todo opcional: el equipo guarda borradores a medias. */
export type TeamPropertyUpdate = Partial<{
  title: string;
  description: string;
  property_type: PropertyType;
  address_line: string;
  city: string;
  province: string;
  postal_code: string;
  neighborhood: string | null;
  latitude: number | null;
  longitude: number | null;
  surface_m2: number | null;
  bedrooms: number;
  bathrooms: number;
  floor: string | null;
  has_elevator: boolean;
  furnished: boolean;
  max_tenants: number;
  total_monthly_rent: number | null;
  deposit: number | null;
  utilities_included: boolean;
  available_from: string | null;
  minimum_stay_months: number | null;
  pets_allowed: boolean | null;
  smoking_allowed: boolean | null;
  couples_allowed: boolean | null;
  students_allowed: boolean | null;
  registration_allowed: boolean | null;
  additional_requirements: string | null;
  amenity_ids: number[];
}>;

export interface NewOwnerInput {
  first_name: string | null;
  last_name: string | null;
  email: string | null;
  phone: string | null;
  owner_type: OwnerType;
  company_name: string | null;
}

export interface AssistedListingCreate {
  owner?: NewOwnerInput;
  owner_profile_id?: number;
  property: TeamPropertyUpdate;
  owner_consent: boolean;
}

export interface AssistedListingResult {
  property_id: number;
  owner_profile_id: number;
  owner_display_name: string;
  owner_email: string;
  is_new_owner: boolean;
  claim_url: string | null;
}
